velox/experimental/cudf/exec/Utilities.cpp

```diff
@@ -0,0 +1,143 @@
+ */
+
+#include "velox/experimental/cudf/exec/Utilities.h"
+
+#include <cudf/concatenate.hpp>
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/mr/device/arena_memory_resource.hpp>
+#include <rmm/mr/device/cuda_async_memory_resource.hpp>
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/managed_memory_resource.hpp>
+#include <rmm/mr/device/owning_wrapper.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+#include <common/base/Exceptions.h>
+
+#include <cstdlib>
+#include <memory>
+#include <string_view>
+
+namespace facebook::velox::cudf_velox {
+
+namespace {
+[[nodiscard]] auto makeCudaMr() {
+  return std::make_shared<rmm::mr::cuda_memory_resource>();
+}
+
+[[nodiscard]] auto makePoolMr() {
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(
+      makeCudaMr(), rmm::percent_of_free_device_memory(50));
+}
+
+[[nodiscard]] auto makeAsyncMr() {
+  return std::make_shared<rmm::mr::cuda_async_memory_resource>();
+}
+
+[[nodiscard]] auto makeManagedMr() {
+  return std::make_shared<rmm::mr::managed_memory_resource>();
+}
+
+[[nodiscard]] auto makeArenaMr() {
+  return rmm::mr::make_owning_wrapper<rmm::mr::arena_memory_resource>(
+      makeCudaMr());
+}
+
+[[nodiscard]] auto makeManagedPoolMr() {
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(
+      makeManagedMr(), rmm::percent_of_free_device_memory(50));
+}
+} // namespace
+
+std::shared_ptr<rmm::mr::device_memory_resource> createMemoryResource(
+    std::string_view mode) {
+  if (mode == "cuda")
+    return makeCudaMr();
+  if (mode == "pool")
+    return makePoolMr();
+  if (mode == "async")
+    return makeAsyncMr();
+  if (mode == "arena")
+    return makeArenaMr();
+  if (mode == "managed")
+    return makeManagedMr();
+  if (mode == "managed_pool")
+    return makeManagedPoolMr();
+  VELOX_FAIL(
+      "Unknown memory resource mode: " + std::string(mode) +
+      "\nExpecting: cuda, pool, async, arena, managed, or managed_pool");
+}
+
+cudf::detail::cuda_stream_pool& cudfGlobalStreamPool() {
+  return cudf::detail::global_cuda_stream_pool();
+};
+
+std::unique_ptr<cudf::table> concatenateTables(
+    std::vector<std::unique_ptr<cudf::table>> tables,
+    rmm::cuda_stream_view stream) {
+  // Check for empty vector
+  VELOX_CHECK_GT(tables.size(), 0);
+
+  if (tables.size() == 1) {
+    return std::move(tables[0]);
+  }
+  std::vector<cudf::table_view> tableViews;
+  tableViews.reserve(tables.size());
+  std::transform(
+      tables.begin(),
+      tables.end(),
+      std::back_inserter(tableViews),
+      [&](const auto& tbl) { return tbl->view(); });
+  return cudf::concatenate(
+      tableViews, stream, cudf::get_current_device_resource_ref());
+}
+
+std::unique_ptr<cudf::table> getConcatenatedTable(
```

Comment:
What if the number of ConcatenatedTable rows is beyond the range? We have a case that accumulated rows in OrderBy is beyond vector_size_t range, https://github.com/facebookincubator/velox/pull/10848

Reply:
@jinchengchenghh how does velox handle larger than `vector_size_t` range for Orderby?

Reply:
velox stores the input in RowContainer, and store the sortedRows pointer in std::vector, so I assume the input rows it can sort is size_t, and extract the outputRows with data type vector_size_t, it will output several batches whose size should satisfy the output rows config.